### PR TITLE
feat(chain): Chunk Only Producers

### PR DESCRIPTION
**Pull request by birchmd**
Chunk-only producers are an important stepping stone towards sharding in mainnet. See https://gov.near.org/t/block-and-chunk-producer-selection-algorithm-in-simple-nightshade/66 for more details. Also see https://github.com/near/NEPs/pull/167 for the spec this work is based on.

This PR does most of the work towards landing this feature. Much of the work with this PR was updating tons of tests because they were using the assumption that validators produce blocks/chunk in a cyclic order. That is no longer true because the randomness is done on the fly at each height instead of when processing the proposals.

This PR is not yet suitable for merging to master; missing items are listed below:

- [ ] Nayduck failures (looks like some tests are failing -- http://nayduck.eastus.cloudapp.azure.com:3000/#/run/1452)
- [ ] Writing a new pytest to see this feature working end-to-end. This PR adds some new tests, and fixes a lot of old tests, so it probably works, but it's always nice to see an integration test.

List of (possible) Nayduck failures to be addressed:
```
expensive nearcore test_rejoin test::test_4_20_kill1_two_shards	
pytest sanity/one_val.py	
pytest sanity/rpc_state_changes.py	
pytest sanity/staking2.py	
pytest sanity/staking_repro1.py	
pytest sanity/state_sync2.py	
pytest sanity/sync_chunks_from_archival.py	
pytest stress/stress.py 3 3 3 0 staking transactions local_network packets_drop	
pytest stress/stress.py 3 3 3 0 staking transactions node_restart packets_drop	
pytest stress/stress.py 3 3 3 0 staking transactions node_restart wipe_data	
pytest sanity/gc_after_sync.py
pytest sanity/gc_sync_after_sync.py swap_nodes
```

### CODE REVIEW

**Comment by bowenwang1996** [comment]
Nayduck run http://nayduck.eastus.cloudapp.azure.com:3000/#/run/2015

File: chain/epoch_manager/src/shard_assignment.rs
Comment (SkidanovAlex) [inline chain/epoch_manager/src/shard_assignment.rs:20]: This is a very confusing name. `shard_validators_heap`, or something along the lines?

File: chain/epoch_manager/src/shard_assignment.rs
Comment (SkidanovAlex) [inline chain/epoch_manager/src/shard_assignment.rs:14]: Are `chunk_producers` sorted in any particular way (in particular, in descending order of stake)?
It would be good to document here.
If they are not sorted in descending stake, it might be worth doing?

File: chain/epoch_manager/src/shard_assignment.rs
Comment (SkidanovAlex) [inline chain/epoch_manager/src/shard_assignment.rs:114]: Between here, and two similar places above, you use inconsistent naming for the first two values you pop (you arbitrarily use or not use the word `least`).

File: chain/epoch_manager/src/shard_assignment.rs
Comment (SkidanovAlex) [inline chain/epoch_manager/src/shard_assignment.rs:95]: This method seems overly complex.

```
random_shuffle(validators) // optionally
last_ord = 0
for (shard_id in ...) {
    for validator_ord in 0..min_validators_per_shard {
        result[shard_id].push(last_ord);
        last_ord += 1
        last_ord %= num_chunk_producers;
    }

}
```

**Review by SkidanovAlex — COMMENTED** [review]
Only reviewed the first half so far, some comments inline.

File: core/primitives/src/rand.rs
Comment (SkidanovAlex) [inline core/primitives/src/rand.rs:161]: This test fails for [4, 6, 1, 1] (with asserts properly updated).

If you look at the aliases section in the page you linked (https://www.keithschwarz.com/darts-dice-coins/), this is exactly the example on which he explains it, but with 6 and 4 swapped. If I don't swap them (leave as 6, 4, 1, 1), your code works correctly (it 6,4,1,1 -> 4,4,3,1 -> 3,4,3,2 -> 3,3,3,3), but if I do, you hit the issue he explains at the very beginning of the section (4,6,1,1 -> 3,6,1,2 -> 3,4,3,2 -> ???).

`big` needs to be a maxheap.

Also, the test itself is extremely flacky (fails with relative error 0.7%-0.8% one out of 5 executions). Though I ran it with `rand::thread_rng` instead of hashes, may be it changes the distribution.

File: chain/epoch_manager/src/validator_selection.rs
Comment (SkidanovAlex) [inline chain/epoch_manager/src/validator_selection.rs:65]: This comment is very confusing. I can't tell if it's intended to be such and I just can't parse it, or if it has something misspelled :)

**Review by SkidanovAlex — COMMENTED** [review]
Still haven't reviewed `validator_selection.rs`.

File: chain/chain/src/store_validator/validate.rs
Comment (bowenwang1996) [reply chain/chain/src/store_validator/validate.rs:681]: That is the difficult part -- there is a test that fails because the catch up has not finished downloading the state and as a result we don't actually have chunk extra for the block. There is nothing wrong with that and I am not sure what would be a better condition to check

File: chain/client/src/client.rs
Comment (bowenwang1996) [reply chain/client/src/client.rs:1428]: > looking at the place where this is called in process_tx_internal it appears that if an active validator received a tx directly, the tx will never be forwarded before this change, resulting in very long processing times

Yes that is exactly the intention. One python test failed because of this.

> But if this is indeed your intent here, the method needs to be renamed?

What do you suggest we rename it to?

**Review by bowenwang1996 — APPROVED** [review]
Approve to unblock this PR and avoid having to deal with merge conflicts with master. If there are any further testing needed, we can do it after it is merged as a nightly protocol feature

**Comment by bowenwang1996** [comment]
@frol looks like it needs your approval 🙏

**Review by frol — APPROVED** [review]
The changes in Rosetta-RPC and primitives LGTM

**Comment by bowenwang1996** [comment]
@nikurt please review

**Review by nikurt — APPROVED** [review]

**Pull request merged**
